---
title: "scafari Vignette"
output: 
    BiocStyle::html_document:
    toc: true
    toc_depth: 2
VignetteBuilder: knitr
Suggests: BiocStyle, knitr, rmarkdown
vignette: >
    %\VignetteIndexEntry{ClEvR Viz vignette}
    %\VignetteEncoding{UTF-8}  
    %\VignetteEngine{knitr::rmarkdown}
editor_options: 
    markdown: 
    wrap: 80
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
Single-cell DNA sequencing (scDNA-seq) enables the analysis of mutation profiles at single-cell resolution. This cutting-edge technique makes it possible to shed light on the previously inaccessible complexity of heterogeneous tissues such as tumors. Even small cell populations that cannot be detected by mass sequencing can now be identified.

## Requirements
To run scafari, you need R (Version 4.4 or higher).

## Installation
## Running scafari
After installation, the package can be loaded into R by typing

```{r loading }
#library(scafari)
```

scafari is available as shiny GUI. You can run the shiny version either by RStudio executing the function `TODO()`. Additionally, all functions for classical use in R are available. The functions can be split into three main/four areas

### Sequencing information and Control?
1. `load` TODO
2. `loglogPlot()`:

### Panel analysis
`normalizeReadCounts()`:
2. `TODO annotate Amp`:
3. `plotAmpliconDistribution()`:
4. `plotAmpliconPerformance()`:
5. `plotPanelUniformity()`:

### Variant analysis
1. `filterVariants()`:
2. `annotateVariants()`:
3. `plotVariantHeatmap()`:
4. `plotGenotypeDistributionPie()`: TODO evtl. Seite 1
5. `plotGenotypeQualityPerGenotype()`:

### Analysis of variants of interest
1. `clusterVariants()`:


# Examplary use of scafari
## Sequencing
The analysis starts by loading the `.h5` file. 

Sequencing Information can be accessed by ...
`sce TODO metadata call`

Read counts per barcode can be investigated using `logLogPlot()`.

## Panel analysis
For panel analysis the read counts per amplicon have to be normalized and annotated with biological relevant information like the corresponding gene names and exons. In the shiny app this is happening in the background. Using the standard R this can be done using `normalizeReadCounts()` and `TODO annotate Amp`.
The locations of the amplicons over the genome can be investigated by visualizing it using `plotAmpliconDistribution()`.
The normalized read counts per amplicons can be assessed in the amplicon performance plot created by `plotAmpliconPerformance()`. 

`normalizeReadCounts()`
`TODO annotate Amp`.
3. `plotAmpliconDistribution()`:
4. `plotAmpliconPerformance()`:
5. `plotPanelUniformity()`: 

## function xxx()
### Details
### Usage
### Examples

## Data
# Session information

```{r session}
sessionInfo()
```